### ISSUE TITLE
squashfs check fails

### ISSUE THREAD

**Issue by sylviette** [kamilion/customizer#1]
Hi,

using python version (4.1.1) as soon as I try to rebuild system, when it starts to check squashfs after having built it, it gets an error and then hangs. It gets an error too when I try to close QEMU, but nevertheless, in this case, it shows desktop and I can work on it anyway.
Notice that it happened with a 4 Gb iso, normally produced by Gambas3 version.
Below follows shell output, taken launching customizer via terminal.

Cheers,
Silvia

---  [shell transcription - build iso]

 silvia@antigone:~$ gksu pkexec /usr/sbin/customizer-gui
* Read Configuration file: /etc/customizer.conf
* Read Configuration file: /etc/customizer.conf
* Read Configuration file: /etc/customizer.conf
* Read Configuration file: /etc/customizer.conf
  => Checking
  => Doing sanity checks
  => Gathering information
  => Cleaning up
  => Updating initramfs
update-initramfs: Generating /boot/initrd.img-4.4.0-11-generic
  => Copying boot files
  => Extracting casper UUID
  => SquashFS Compression parameters: ('-b', '1048576', '-comp', 'gzip')
  => Creating SquashFS Compressed Filesystem
Parallel mksquashfs: Using 4 processors
Creating 4.0 filesystem on /home/ISO/casper/filesystem.squashfs, block size 1048576.
[======================================================================================================================================================-] 414311/414311 100%

Exportable Squashfs 4.0 filesystem, gzip compressed, data block size 1048576
	compressed data, compressed metadata, compressed fragments, compressed xattrs
	duplicates are removed
Filesystem size 4420000.05 Kbytes (4316.41 Mbytes)
	38.27% of uncompressed filesystem size (11550707.40 Kbytes)
Inode table size 6275553 bytes (6128.47 Kbytes)
	25.95% of uncompressed inode table size (24180776 bytes)
Directory table size 6345649 bytes (6196.92 Kbytes)
	38.38% of uncompressed directory table size (16535520 bytes)
Number of duplicate files found 128507
Number of inodes 627013
Number of files 416270
Number of fragments 5502
Number of symbolic links  173291
Number of device nodes 0
Number of fifo nodes 0
Number of socket nodes 0
Number of directories 37452
Number of ids (unique uids + gids) 34
Number of uids 14
	root (0)
	... 
[UIDS/GIDS LIST HERE]
  => Checking SquashFS filesystem size

[here hangs without producing iso]

---  [shell transcription - QEMU]

silvia@antigone:~$ gksu pkexec /usr/sbin/customizer-gui
* Read Configuration file: /etc/customizer.conf
* Read Configuration file: /etc/customizer.conf
* Read Configuration file: /etc/customizer.conf
* Read Configuration file: /etc/customizer.conf

  => Checking
  => Detecting available sessions
  => Starting Xephyr
  => Allowing local access to X-server
non-network local connections being added to access control list
  => Starting nested X session: /usr/lib/gnome-flashback/gnome-flashback-metacity
GNOME_KEYRING_CONTROL=/etc/skel/.cache/keyring-MWFFEY
GNOME_KEYRING_CONTROL=/etc/skel/.cache/keyring-MWFFEY
SSH_AUTH_SOCK=/etc/skel/.cache/keyring-MWFFEY/ssh
GNOME_KEYRING_CONTROL=/etc/skel/.cache/keyring-MWFFEY
SSH_AUTH_SOCK=/etc/skel/.cache/keyring-MWFFEY/ssh
Fcitx seems is not running
  => Terminating Xephyr
[Error here]
[dix] Could not init font path element /usr/share/fonts/X11/cyrillic, removing from list!
[dix] Could not init font path element /usr/share/fonts/X11/100dpi/:unscaled, removing from list!
[dix] Could not init font path element /usr/share/fonts/X11/75dpi/:unscaled, removing from list!
[dix] Could not init font path element /usr/share/fonts/X11/100dpi, removing from list!
[dix] Could not init font path element /usr/share/fonts/X11/75dpi, removing from list!

**Comment by kamilion** [comment]
Hm, is it reproducible, or is this the only instance you've seen the error so far?

**Comment by sylviette** [comment]
Hi Kamilion,
yes, I've reproduced it right now but Ive noticed that it happens only with 16.04 (updated today, so beta1 pre-UI-freeze); I've tried with trusty too and everything works fine, while with xenial it hangs when checking SquashFS filesystem size as it did before.

About trusty, btw, there's a problem anyway: when trying to load iso, it hanged with an error warning (too small and not realizable to see anything else but "critical"). I don't report anything, this time, because on shell is just visible what was OK with no error when hanging.

Notice that Gambas3 version works well for loading new isos (both trusty and xenial) and for rebuilding them, while it gives tons of errors in installation process via shell (xenial only); so now I'm loading xenial iso with Gambas3 version, then managing it with Python one and finally rebuilding iso with Gambas3 again.

In these days I'll try to install xenial on a VM and run Customizer on it: maybe it depends on working in a older Ubuntu version on a newer one. I'll let you know.

Thanks and cheers,
Silvia

**Comment by sylviette** [comment]
I've found sort of a dirty workaround, good for Xenial too, but with one important limitation.
Briefly, I load xenial iso using gambas version, then remaster it with py version and finally build it with gb again.

The limitation I was talking about before has to do - I can imagine - with mkisofs that doesn't allow files to be larger than 4 GB; since I'll have to upgrade from trusty to Xenial (when it will be released and tested) a multilanguage distribution (18 langs and tons of software) that will weigh, I fear, more than 4 Gb, I'm wondering if there's a way to solve/bypass mkisofs 4Gb limitation or if the only way would be to shrink it in some way, eliminating something?

Thanks,
Silvia

**Comment by sylviette** [comment]
Hi Kamilion,

I've found sort of a dirty workaround, good for Xenial too, but with one important limitation.
Briefly, I load xenial iso using gambas version, then remaster it with py version and finally build it with gb again.

The limitation I was talking about before has to do - I can imagine - with mkisofs that doesn't allow files to be larger than 4 GB; since I'll have to upgrade from trusty to Xenial (when it will be released and tested) a multilanguage distribution (18 langs and tons of software) that will weigh, I fear, more than 4 Gb, I'm wondering if there's a way to solve/bypass mkisofs 4Gb limitation or if the only way would be to shrink it in some way, eliminating something?

Thanks,
Silvia

**Comment by kamilion** [comment]
... whoa.... 4GB limitation? How big is the ISO you've built?!

I don't think there should be any limitation in ISO9660 of total volume size, but there may be a maximum size of any single file... Ah, squashfs file is bigger than 4GB. There's ways you can get around that, actually.

https://github.com/kamilion/kamikazi-core/tree/master/resources/buildscripts/xenial

I have here some buildscripts that I make my own ISOs from.

Ubuntu's livecd script, Casper, is actually quite flexible.

https://github.com/kamilion/kamikazi-core/tree/master/resources/xenial/mods/usr/share/initramfs-tools/scripts/casper-bottom

Using a script like I have, in the casper-bottom directory, you could probably figure out how to union mount as many 4GB squashfs files as you want during early startup of the ISO.

If you're doing something that has a huge /srv or /opt for instance, you can split /srv or /opt out into their own squashfs files.

Or if you're including a huge homedir or something...


The only real solution to breaking the 4GB filesize I can think of is passing -udf to 'xorriso -as mkisofs', which isn't going to work with xorriso, considering I'm pretty sure it doesn't support the UDF filesystem according to https://www.gnu.org/software/xorriso/


I'll have a look around, but right now I'd suggest shrinking your squashfs below the 4GB size limit.
Drop big documentation packages like qt4-doc in your build process.
```
# Purge the qt4-doc package, it's a hundred megs of incompressableness.
apt-get purge -y qt4-doc
```

https://github.com/kamilion/kamikazi-core/blob/master/resources/buildscripts/xenial/00-build-clean-iso-from-source.sh#L59

**Comment by kamilion** [comment]
No response in 5 months; closing issue as 'wontfix' due to the underlying 4GB ISO limitation being the problem here, I'll keep on the look out for potential fixes using UDF if something changes in the ISO generation tools to make that easy to access. If so; I'll probably put an automatic check for an oversize squashfs to switch to UDF and no user change to workflow should be required.

**Comment by sylviette** [comment]
Hi Kamilion,

sorry for my absence due to personal/family-related problems.

Since your workaround seems to be a little long for time and resources I can give to our students' iso project, I've decided to follow your suggestion and uninstalled QTDeveloper: iso target is about 32K students and several staff persons and how many of them will develop in C++ with QTDeveloper? I've removed QTDeveloper itself saving about 1.2 Gb and created a textfile on /etc/skel/Desktop with links and instructions for people who's interested in ...

Dirty trick but efficient and fast.

Thanks again,
Silvia
